provider/aws: add support for geolocation and latency records to route53

acm1 opened the pull request:
This should fulfill the feature request in https://github.com/hashicorp/terraform/issues/2746.

robzienert commented:
:+1:

apparentlymart commented on builtin/providers/aws/resource_aws_route53_record.go:
```diff
@@ -94,6 +94,35 @@ func resourceAwsRoute53Record() *schema.Resource {
 				Optional: true,
 			},
 
+			"region": &schema.Schema{ // AWS region from which to evaluate latency
```
The usual convention is to have the Terraform properties match with the AWS API or UI labels for things, but this name seems rather confusing and I think you agree since you put a comment on it trying to elaborate. :grinning:

Perhaps this should be called ``latency_routing_region``. My intent with this name is to indicate that we're giving the region that will be used for latency routing, and not (as would normally be the case where "region" is used elsewhere in Terraform) where this record resource will be hosted.

Alternatively, riffing off how this is arranged in the Route53 UI, I might consider having a ``routing_policy`` child block that contains all of these routing policy options. This is similar to how it's placed in a separate box in the Route53 console, distinct from the TTL and value.

```js
resource "aws_route53_record" "foo" {
    zone_id = "${aws_route53_zone.primary.zone_id}"
    name = "www.example.com"
    type = "A"
    ttl = "300"
    records = ["${aws_eip.lb.public_ip}"]

    routing_policy {
        region = "us-west-2"
    }
}
```

Then at least it's a bit clearer that this region is routing-related. Your other set of mutually-exclusive fields 

That last alternative still misses the "routing policy" field from the UI where you choose a type of routing, indicating that this is latency-based routing, so maybe this is better:

```js
resource "aws_route53_record" "foo" {
    zone_id = "${aws_route53_zone.primary.zone_id}"
    name = "www.example.com"
    type = "A"
    ttl = "300"
    records = ["${aws_eip.lb.public_ip}"]

    latency_routing_policy {
        region = "us-west-2"
    }
}
```

...and then there could be other mutually-exclusive child blocks for ``weighted_routing_policy``, ``failover_routing_policy`` and ``geolocation_routing_policy``, each containing the set of fields that gets displayed when you select the corresponding option for routing policy in the Route53 console.

Not sure which of these options I like the best. I think I lean towards the latter since I think it's most user-friendly to prefer to mimic the AWS UI rather than the API when they are in conflict, but YMMV.

apparentlymart commented:
Looking forward to having these features available in Terraform! Thanks for doing this.

If you've got the time it'd be nice to also update the website docs for this resource, which you can find in ``website/source/docs/providers/aws/r/route53_record.html.markdown``.

catsby commented:
I agree with @apparentlymart regarding the naming of the `region` attribute, what do you think the suggestions made, @acm1 ?

farridav commented:
+1 Thanks for putting this up, keen to see it in.. as Ill be needing this feature very soon

farridav commented:
*bump .. any movement on this @acm1 ?  ... Thanks

acm1 commented:
I'll take a look at getting @apparentlymart's suggestions into the PR this week.

farridav commented:
Awesome, thanks @acm1

shanielh commented:
Looking forward for this one to be merged

catsby commented:
Hey @acm1 any chance you're still working on this?

catsby commented:
Another follow up here, sorry...

These additions look very promising and there seems to be good desire from the community to have, so thank you for the start! 

In order to merge this PR I'd like to see the following:

- [ ] Change to the block style that @apparentlymart had suggested 
- [ ] Acceptance tests for the behavior
- [ ] Documentation for the new attributes

If you have questions on how to convert to the block style, let us know! You should be able to use Route53's `alias` attribute as a reference. 

For acceptance tests, you can probably use the same `alias`'s test as an example:

- https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_route53_record_test.go#L194

For documentation, look in the `website` folder in the root of the Terraform project, and the README will help get you started on running the site locally. From there, it's adding documentation in Markdown format

- https://github.com/hashicorp/terraform/tree/master/website

Let us know if you have other questions!

acm1 commented:
This one got away from me. I'll dig into it this week -- thanks for the guidance @catsby.

catsby commented:
Great to hear @acm1 , thanks! Let me know if you have questions

acm1 commented:
@catsby I've pushed changes to implement @apparentlymart's suggested block style along with acceptance tests and documentation. Let me know what you think.

TheLinuxNinja commented:
Has this been released, yet?

catsby commented:
@TheLinuxNinja no. I recently refactored Route53 Records internals a bit so this will need a rebase (sorry about that, but [it was necessary][1])

Let me know if you (@acm1) will be able to do that, thanks!

[1]: https://github.com/hashicorp/terraform/pull/4892

acm1 commented:
@catsby I've rebased.

catsby commented:
excellent, thanks! the Travis error is a thing being worked out, so I should merge this early next week

acm1 commented on builtin/providers/aws/resource_aws_route53_record.go:
```diff
@@ -507,12 +635,12 @@ func expandRecordName(name, zone string) string {
-	buf.WriteString(fmt.Sprintf("%s-", m["zone_id"].(string)))
-	buf.WriteString(fmt.Sprintf("%t-", m["evaluate_target_health"].(bool)))
-
-	return hashcode.String(buf.String())
```
Removed `resourceAwsRoute53AliasRecordHash` because it seems that the default hashing function (schema.HashResource) will handle this fine. I could be reading the code wrong though...

acm1 commented:
Besides my code comment above, I was thinking maybe it makes more sense to put `set_identifier` inside the routing blocks instead of outside. E.g.
```
resource "aws_route53_record" "foo" {
    ...
    latency_routing_policy {
        set_identifier = "Oregon"
        region = "us-west-2"
    }
}
```
vs.
```
resource "aws_route53_record" "foo" {
    ...
    set_identifier = "Oregon"
    latency_routing_policy {
        region = "us-west-2"
    }
}
```
Thoughts?

oli-g commented:
When this is going to be merged?

catsby commented:
Hey all – I'm re-reviewing this now/today. I need to think on some things as this will introduce a backwards incompatibility regarding weighted records that we already have.

@acm1 regarding the `set_identifier`, I think we're OK to leave it out of the block as it is now. My understanding is that it's applicable to all of these policy types, and each individual record can only have one of those types (weighted, latency, etc)

acm1 commented on builtin/providers/aws/resource_aws_route53_record.go:
```diff
@@ -68,13 +66,10 @@ func resourceAwsRoute53Record() *schema.Resource {
 			"weight": &schema.Schema{
 				Type:     schema.TypeInt,
 				Optional: true,
-				Default:  -1,
```
I'm thinking I should leave this line in so that terraform doesn't try to modify every route53 resource already on record since the default value changes from -1 to 0 if this line is removed.

catsby replied:
We can address this in a migration, e.g:

- https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/resource_aws_security_group_rule_migrate.go

We'll need to do that for all `-1` entries.

catsby commented on builtin/providers/aws/resource_aws_route53_record.go:
```diff
@@ -264,14 +347,23 @@ func resourceAwsRoute53RecordRead(d *schema.ResourceData, meta interface{}) erro
+		d.Set("geolocation_routing_policy.subdivision", record.GeoLocation.SubdivisionCode)
+	}
+	if record.Region != nil {
+		d.Set("latency_routing_policy.region", record.Region)
```
There are issues here, in that we are not checking the error returned from `d.Set` here. For simple values of `TypeInt` or `TypeString` this is OK, however for complicated things like `TypeSet` or `TypeList` we need to check for errors. 

Here I know we're erring, as you can't do a deep set like this (`geolocation_routing_policy.subdivision`) you must set the entire `geolocation_routing_policy` set. As a result we're seeing a perpetual plan here.

romansky commented:
Any news on this front?

oli-g commented:
I think a lot of people needs this feature!

simple-guy commented:
+1

acm1 commented:
@catsby I've reworked this a bit and addressed your comments by adding a state migration and checking for errors when setting existing values. Thoughts?

gechr commented:
Just compiled/tested this and can confirm the new `geolocation_routing_policy` works as advertised.

@catsby From your point of view, what's left here before this can be merged? Keen to move back to an upstream branch, rather relying on a fork! 😄

catsby commented:
Hey all – I'm going to continue this work in https://github.com/hashicorp/terraform/pull/6954 ! Thank you @acm1 !

catsby commented:
#6954 was just merged! Thank you @acm1 and I sincerely apologize for the long, long, long delay in getting this in 😄
